src/org/omegat/gui/editor/filter/ReplaceFilter.java

```diff
@@ -107,13 +110,23 @@ public void replaceAll() {
             if (trans == null) {
                 continue;
             }
+            // Avoid to replace more than once with variables when entries have duplicates
+            if ((en.defaultTranslation) && (ste.getDuplicate() == SourceTextEntry.DUPLICATE.NEXT)) {
+                continue; // Already replaced when we parsed the first entry
+            }
             List<SearchMatch> found = getReplacementsForEntry(trans);
             if (found != null) {
                 int off = 0;
                 StringBuilder o = new StringBuilder(trans);
                 for (SearchMatch m : found) {
-                    o.replace(m.getStart() + off, m.getEnd() + off, replacement);
-                    off += replacement.length() - m.getLength();
+                    String repl;
+                    try {
+                        repl = ((ReplaceMatch) m).getReplacement();
```

Comment:
Not only to impact less code. Another reason is to avoid to repeat an identical string in all replace match instances: the only case where it can differ is when you use regex replacements.
That is also one other reason for using a special ReplaceMatch class: this class is susceptible to have thousands of instances for a single search, while only a few of them really need to have a string distinct from the original one.

Reply:
Let's move this discussion down to a separate thread.